controllers/helpers.go

```diff
@@ -0,0 +1,77 @@
+package controllers
+
+import (
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	upgrademgrv1alpha1 "github.com/keikoproj/upgrade-manager/api/v1alpha1"
+	"log"
+	"strconv"
+	"strings"
+)
+
+// getMaxUnavailable calculates and returns the maximum unavailable nodes
+// takes an update strategy and total number of nodes as input
+func getMaxUnavailable(strategy upgrademgrv1alpha1.UpdateStrategy, totalNodes int) int {
+	// Below are the constants set in intstr package
+	// const (
+	//	Int    Type = iota // The IntOrString holds an int.
+	//	String             // The IntOrString holds a string.
+	//)
+	maxUnavailable := 1
+	if strategy.MaxUnavailable.Type == 0 {
+		maxUnavailable = int(strategy.MaxUnavailable.IntVal)
+	} else if strategy.MaxUnavailable.Type == 1 {
+		strVallue := strategy.MaxUnavailable.StrVal
+		intValue, _ := strconv.Atoi(strings.Trim(strVallue, "%"))
+		maxUnavailable = int(float32(intValue) / float32(100) * float32(totalNodes))
+	}
+	// setting maxUnavailable to total number of nodes when maxUnavailable is greater than total node count
+	if totalNodes < maxUnavailable {
+		log.Printf("Reducing maxUnavailable count from %d to %d as total nodes count is %d",
+			maxUnavailable, totalNodes, totalNodes)
+		maxUnavailable = totalNodes
+	}
+	// maxUnavailable has to be atleast 1 when there are nodes in the ASG
+	if totalNodes > 0 && maxUnavailable < 1 {
+		maxUnavailable = 1
+	}
```

Comment:
It might be worth logging what the computed value of maxUnavailable is. This would help debugging.

Reply:
done